src/main/java/org/jitsi/videobridge/RtpChannel.java

```diff
@@ -1991,6 +1968,121 @@ public RtpChannelTransformEngine getTransformEngine()
+                    && sgpe.getSources() != null
+                    && sgpe.getSources().size() >= 2)
+                {
+                    // Every simulcast group determines an mst with a bunch of
+                    // encodings.
+                    MediaStreamTrack track = new MediaStreamTrack();
+
+                    int order = RTPEncoding.BASE_ORDER;
+                    for (SourcePacketExtension spe : sgpe.getSources())
+                    {
+                        long primarySSRC = spe.getSSRC();
+                        freeSSRCs.remove(primarySSRC);
+
+                        long[] encoding = new long[] { -1, -1, -1 };
+                        encoding[0] = primarySSRC;
+                        encoding[1] = order++;
+
+                        encodings.put(primarySSRC, encoding);
+                        tracksBySSRC.put(primarySSRC, track);
+                    }
+                }
+                else if ("fid".equalsIgnoreCase(sgpe.getSemantics())
+                    && sgpe.getSources() != null
+                    && sgpe.getSources().size() == 2)
+                {
+                    Long primarySSRC = sgpe.getSources().get(0).getSSRC();
+                    freeSSRCs.remove(primarySSRC);
+                    Long rtxSSRC = sgpe.getSources().get(1).getSSRC();
+                    freeSSRCs.remove(rtxSSRC);
+
+                    long[] encoding = encodings.get(primarySSRC);
+                    if (encoding == null)
+                    {
+                        // if we don't have an encoding, then we don't have
+                        // an mst. create both.
+                        MediaStreamTrack track = new MediaStreamTrack();
+                        tracksBySSRC.put(primarySSRC, track);
+
+                        encoding = new long[] { -1, -1, -1 };
+                        encoding[0] = primarySSRC;
+                        encoding[1] = RTPEncoding.BASE_ORDER;
+
+                        encodings.put(primarySSRC, encoding);
+                        tracksBySSRC.put(primarySSRC, track);
+                    }
+
+                    encoding[2] = rtxSSRC;
+                }
+            }
+        }
+
+        for (Map.Entry<Long, MediaStreamTrack> entry : tracksBySSRC.entrySet())
+        {
+            MediaStreamTrack track = entry.getValue();
+            long[] encoding = encodings.get(entry.getKey());
+            track.addEncoding(encoding[0], encoding[2], -1, (int) encoding[1]);
+        }
+
+        Map<Long, MediaStreamTrack> remoteTracks = stream.getRemoteTracks();
+
+        synchronized (remoteTracks)
+        {
+            remoteTracks.clear();
+            remoteTracks.putAll(tracksBySSRC);
+
+            for (Long primarySSRC : freeSSRCs)
+            {
+                MediaStreamTrack mst = new MediaStreamTrack();
+                mst.addEncoding(primarySSRC, -1, -1, RTPEncoding.BASE_ORDER);
+                remoteTracks.put(primarySSRC, mst);
```

Comment:
That's murky business. I saw that you exposed the private field/state of `MediaStreamImpl` through `#getRemoteTracks()` and I let it be because I didn't want to stop your libjisti PR. However, this is a bit too much for me to swallow without mentioning it at least. You get the Map from `stream` without being 100% sure as a client that the implementation is not a copy of the private state, then you synchronize on it, and then you go an modify it hoping that you're modifying the private state of `MediaStreamImpl`!